src/lib.rs

```diff
@@ -195,38 +210,67 @@ where
     Ok(result)
 }
 
+fn close(event_loop: &PyAny) -> PyResult<()> {
+    event_loop.call_method1(
+        "run_until_complete",
+        (event_loop.call_method0("shutdown_asyncgens")?,),
+    )?;
+
+    // how to do this prior to 3.9?
+    if event_loop.hasattr("shutdown_default_executor")? {
+        event_loop.call_method1(
+            "run_until_complete",
+            (event_loop.call_method0("shutdown_default_executor")?,),
+        )?;
+    }
+
+    event_loop.call_method0("close")?;
+
+    Ok(())
+}
+
 /// Attempt to initialize the Python and Rust event loops
 ///
 /// - Must be called before any other pyo3-asyncio functions.
 /// - Calling `try_init` a second time returns `Ok(())` and does nothing.
 ///   > In future versions this may return an `Err`.
+#[deprecated(since = "0.14.0")]
 pub fn try_init(py: Python) -> PyResult<()> {
-    EVENT_LOOP.get_or_try_init(|| -> PyResult<PyObject> {
-        let asyncio = py.import("asyncio")?;
-        let ensure_future = asyncio.getattr("ensure_future")?;
-        let event_loop = asyncio.call_method0("get_event_loop")?;
+    CACHED_EVENT_LOOP.get_or_try_init(|| -> PyResult<PyObject> {
+        let event_loop = asyncio_get_event_loop(py)?;
         let executor = py
             .import("concurrent.futures.thread")?
-            .getattr("ThreadPoolExecutor")?
-            .call0()?;
+            .call_method0("ThreadPoolExecutor")?;
         event_loop.call_method1("set_default_executor", (executor,))?;
-        let call_soon = event_loop.getattr("call_soon_threadsafe")?;
-        let create_future = event_loop.getattr("create_future")?;
-
-        ASYNCIO.get_or_init(|| asyncio.into());
-        ENSURE_FUTURE.get_or_init(|| ensure_future.into());
-        EXECUTOR.get_or_init(|| executor.into());
-        CALL_SOON.get_or_init(|| call_soon.into());
-        CREATE_FUTURE.get_or_init(|| create_future.into());
+
+        EXECUTOR.set(executor.into()).unwrap();
         Ok(event_loop.into())
     })?;
 
     Ok(())
 }
 
+fn asyncio(py: Python) -> PyResult<&PyAny> {
+    ASYNCIO
+        .get_or_try_init(|| Ok(py.import("asyncio")?.into()))
+        .map(|asyncio| asyncio.as_ref(py))
+}
+
+fn asyncio_get_event_loop(py: Python) -> PyResult<&PyAny> {
+    asyncio(py)?.call_method0("get_event_loop")
+}
+
 /// Get a reference to the Python Event Loop from Rust
+pub fn get_running_loop(py: Python) -> PyResult<&PyAny> {
+    // Ideally should call get_running_loop, but calls get_event_loop for compatibility between
+    // versions.
+    asyncio(py)?.call_method0("get_event_loop")
```

Comment:
`get_running_loop` exists in the `asyncio` API, but it was added in 3.7. Since we want to maintain compatibility with 3.6, we are limited to the functionality it provides. We could potentially try `get_running_loop` first, _then_ try `get_event_loop` for compatibility, but that only fixes it for Python >3.7 (which is probably ok as long as we document it).

I don't think it necessarily needs to be public (although I'm not 100% sure on that). The reason we need it at all is because it's used when Python calls into pyo3. Essentially, when a Python coroutine calls a `#[pyfunction]` from the `asyncio` event loop, we can use `get_running_loop` to get the event loop associated with the current OS thread and store it as a task local.

If we made it private, the same problems would still exist since I believe something like this has to exist internally.

Reply:
> I'm slightly uncomfortable with this behaviour, as anyone can call this coming from the rust-async threads, and then accidentally create an event loop bound to to the worker thread that task was calling it from. 

The docs for `get_event_loop` state that it will only try to create an event loop if the current OS thread is the main thread, so I believe this is not possible. It's true that it would be problematic for people to call this function from any rust thread, but I think the worst it does is error out.

Reply:
I went ahead and made this function call `get_running_loop` if available and fall back to `get_event_loop` if not. This will make the function a bit more permissive on Python 3.6, but the difference is documented now.